Library/Formula/node.rb

```diff
@@ -55,42 +55,52 @@ def install
     system "./configure", *args
     system "make", "install"
 
-    resource("npm").stage libexec/"npm" if build.with? "npm"
-  end
+    if build.with? "npm"
+      resource("npm").stage buildpath/"npm_install"
 
-  def post_install
-    return if build.without? "npm"
+      # make sure npm can find node
+      ENV["PATH"] = "#{Formula["node"].prefix}/bin:#{ENV["PATH"]}"
 
-    (libexec/"npm").cd { system "make", "uninstall" }
-    Pathname.glob(HOMEBREW_PREFIX/"share/man/*") do |man|
-      next unless man.directory?
-      man.children.each do |file|
-        next unless file.symlink?
-        file.unlink if file.readlink.to_s.include? "/node_modules/npm/man/"
+      # set log level temporarily for npm's `make install`
+      ENV["NPM_CONFIG_LOGLEVEL"] = "verbose"
+
+      cd buildpath/"npm_install" do
+        system "./configure", "--prefix=#{libexec}/npm"
+        system "make", "install"
       end
+
+      rm_rf libexec/"npm/lib/node_modules/npm/share"
+
+      if build.with? "completion"
+        bash_completion.install \
+          buildpath/"npm_install/lib/utils/completion.sh" => "npm"
+      end
+
+      # Remove the now-redundant directory.
+      rm_rf buildpath/"npm_install"
     end
+  end
+
+  def post_install
+    return if build.without? "npm"
 
     node_modules = HOMEBREW_PREFIX/"lib/node_modules"
     node_modules.mkpath
-    cp_r libexec/"npm", node_modules
+    npm_exec = node_modules/"npm/bin/npm-cli.js"
+    # Kill npm but preserve all other modules across node updates/upgrades.
+    rm_rf node_modules/"npm"
```

Comment:
Due to ` npm ` shipping a trillion manpages, the code ended up being less, ah, pretty. Any thoughts on compacting it?

```ruby

    # Let's do the manpage dance. It's just a jump to the left.
    # And then a step to the right, with your hand on rm_f.
    rm_f Dir[HOMEBREW_PREFIX/"share/man/man1/{npm.,npm-}*1"]
    rm_f Dir[HOMEBREW_PREFIX/"share/man/man3/{npm.,npm-}*3"]
    rm_f Dir[HOMEBREW_PREFIX/"share/man/man5/{npm.,npm-}*5"]
    rm_f Dir[HOMEBREW_PREFIX/"share/man/man7/{npm.,npm-}*7"]
    # Then install them as expected.
    man1.install Dir[node_modules/"npm/man/man1/*"]
    man3.install Dir[node_modules/"npm/man/man3/*"]
    man5.install Dir[node_modules/"npm/man/man5/*"]
    man7.install Dir[node_modules/"npm/man/man7/*"]
    # And make sure they're linked.
    Dir[man1/"npm*"].each { |f| ln_s f, HOMEBREW_PREFIX/"share/man/man1" }
    Dir[man3/"npm*"].each { |f| ln_s f, HOMEBREW_PREFIX/"share/man/man3" }
    Dir[man5/"npm*"].each { |f| ln_s f, HOMEBREW_PREFIX/"share/man/man5" }
    Dir[man7/"npm*"].each { |f| ln_s f, HOMEBREW_PREFIX/"share/man/man7" }
```